lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -428,93 +433,165 @@ static SILValue emitCodeForSymbolicValue(SymbolicValue symVal,
-/// required. Since this pass folds only integers or strings and since the
-/// former is a trivial type, we only have to destroy strings that are folded.
-/// For strings, a release_value (or a destory_value instruction in ownership
-/// SIL) has to be emitted if it is not already present.
+/// Replace all uses of \c originalVal by \c foldedVal and adjust lifetimes of
+/// original and folded values by emitting required destory/release instructions
+/// at the right places. Note that this function does not remove any
+/// instruction.
+///
+/// \param originalVal the SIL value that is replaced.
+/// \param foldedVal the SIL value that replaces the \c originalVal.
+/// \param fun the SIL function containing the \c foldedVal and \c originalVal
 static void
-destroyFoldedValueAtEndOfUse(SILValue foldedVal, SILValue originalVal,
-                             ArrayRef<SILInstruction *> endOfUseInsts,
-                             SILFunction *fun) {
-  // Folded value should have either trivial or owned ownership as it is an
-  // integer or string constant.
-  assert(foldedVal.getOwnershipKind() == ValueOwnershipKind::None ||
-         foldedVal.getOwnershipKind() == ValueOwnershipKind::Owned);
-
-  // If the ownership kinds of folded and original values are both either
-  // owned or trivial, there is nothing to do.
-  if (foldedVal.getOwnershipKind() == originalVal.getOwnershipKind()) {
+replaceAllUsesAndFixLifetimes(SILValue foldedVal, SILValue originalVal,
+                              SILFunction *fun) {
+  if (originalVal->getType().isTrivial(*fun)) {
+    assert(foldedVal->getType().isTrivial(*fun));
+    originalVal->replaceAllUsesWith(foldedVal);
     return;
   }
-  assert(originalVal.getOwnershipKind() == ValueOwnershipKind::Guaranteed);
+  assert(!foldedVal->getType().isTrivial(*fun));
+
+  // A helper for releasing values in OSSA and non-OSSA.
+  auto releaseValue = [&](SILBuilder &builder, SILLocation loc, SILValue val) {
+    if (fun->hasOwnership()) {
+      builder.emitDestroyValueOperation(loc, val);
+    } else {
+      builder.createReleaseValue(loc, val, builder.getDefaultAtomicity());
+    }
+  };
 
-  // Here, the original value may be at +0 and hence may not be released.
-  // However, the folded value should always be released.
-  SmallVector<SILInstruction *, 2> lifeTimeEndInstsOfOriginal;
-  getLifetimeEndInstructionsOfSILValue(originalVal, lifeTimeEndInstsOfOriginal);
+  // In the non-trival case, the foldedVal must at +1 as it must be a string
+  // constant. The originalVal could be at +0 or +1.
 
-  if (!lifeTimeEndInstsOfOriginal.empty()) {
-    // Here, the original value is released, and so would be the folded value.
-    return;
-  }
+  // Iterate over all non-arc uses of originalVal and replace it with foldedVal,
+  // after emitting required ARC operation. The foldedVal should be at +1 even
+  // after replacing it in all the uses of originalVal.
+  while (Operand *use = getFirstNonARCUse(originalVal)) {
```

Comment:
I guess this whole loop can go away if you limit the pass to OSSA.

Reply:
Yes. I have just pushed a commit that removes this loop